### PR TITLE
LibGUI+Browser: Add Vertical Tabs

### PR DESCRIPTION
**Pull request by cammo1123**
## LibGUI::TabWidget
Add Vertical tabs to GUI::TabWidget, This can be set using one of the two new TabWidget::TabPosition's

```cpp
enum TabWidget::TabPosition { Top, Bottom, Left, Right };
// or by using 
void set_tab_position(TabWidget::TabPosition position);
```
## Browser
Using the new TabLayouts the browser now has a new key combination ``` CTRL + , ``` that switches between a ```Top``` positioned  and a ```Left``` positioned tab layout.

image

### CODE REVIEW

File: Userland/Libraries/LibGfx/ClassicStylePainter.cpp [hunk 1]
```diff
@@ -30,7 +30,12 @@ void ClassicStylePainter::paint_tab_button(Painter& painter, IntRect const& rect
     PainterStateSaver saver(painter);
     painter.translate(rect.location());
 
-    if (top) {
+    // 0: Top
+    // 1: Bottom
+    // 2: Left
+    // 3: Right
+
+    if (position == 0) {
```
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibGfx/ClassicStylePainter.cpp:38]: That's no reason to make the code worse :)

[the remark this answers is not in the document]
Comment (networkException) [reply Userland/Libraries/LibGfx/ClassicStylePainter.cpp:38]: In my opinion that optimization is not worth the worse readability

File: Userland/Libraries/LibGUI/TabWidget.cpp [hunk 2]
```diff
@@ -368,6 +364,31 @@ void TabWidget::set_bar_visible(bool bar_visible)
...
+    } else {
+        rect.translate_by(0, -2);
+        rect.set_height(rect.height() + 4);
+    }
+
+    rect.translate_by(bar_rect().location());
+    return rect;
+}
+
+Gfx::IntRect TabWidget::horz_button_rect(size_t index) const
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/TabWidget.cpp:392]: Don't uselessly abbreviate here. Use `vertical` and `horizontal`.

File: Userland/Libraries/LibGfx/ClassicStylePainter.cpp [hunk 3]
```diff
@@ -17,7 +17,7 @@
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGfx/ClassicStylePainter.cpp:18]: This commit should be squashed with the first one.

File: Userland/Libraries/LibGfx/ClassicStylePainter.cpp [hunk 4]
```diff
@@ -35,7 +35,7 @@ void ClassicStylePainter::paint_tab_button(Painter& painter, IntRect const& rect
     // 2: Left
     // 3: Right
 
-    if (position == 0) {
+    if (position == GUI::TabWidget::TabPosition::Top) {
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGfx/ClassicStylePainter.cpp:34]: It looks like this is a hand-rolled switch case.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Please add the tab position to GML with `REGISTER_ENUM_PROPERTY`.

**Comment by cammo1123** [comment]
Commits have been squished, and all issues pointed out to me have been resolved (as far as I know)
- [x] Added GML support

File: Userland/Libraries/LibGUI/TabWidget.cpp [hunk 5]
```diff
@@ -27,6 +29,11 @@ TabWidget::TabWidget()
 
     REGISTER_MARGINS_PROPERTY("container_margins", container_margins, set_container_margins);
     REGISTER_BOOL_PROPERTY("uniform_tabs", uniform_tabs, set_uniform_tabs);
+    REGISTER_ENUM_PROPERTY("tab_position", this->tab_position, set_tab_position, TabPosition,
+        { TabPosition::Top, "Top" },
+        { TabPosition::Bottom, "Bottom" },
+        { TabPosition::Left, "Left" },
+        { TabPosition::Right, "Right" }, );
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/TabWidget.cpp:36]: Can you also add this property to the table in https://github.com/SerenityOS/serenity/blob/master/Base/usr/share/man/man5/GML-Widget-TabWidget.md?

File: Userland/Libraries/LibGUI/TabWidget.cpp [hunk 6]
```diff
@@ -1,11 +1,13 @@
 /*
  * Copyright (c) 2018-2021, Andreas Kling <kling@serenityos.org>
  * Copyright (c) 2021, Peter Elliott <pelliott@serenityos.org>
+ * Copyright (c) 2022, Cameron Youell <cameronyouell@gmail.com>
  * Copyright (c) 2022, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include "LibCore/Object.h"
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/TabWidget.cpp:10]: Angle bracket includes, please

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by cammo1123** [comment]
Sorry for all the force commits. I'm not too familiar with git, and I had to resolve some conflicts. It should be all up to date and conflict-free now!

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by AtkinsSJ** [comment]
You've ended up with a merge commit in there. You can remove it with `git rebase -i master` and then deleting the line with that commit. (Make sure you don't delete anything else though! But if you do, you can rescue it still, ask on Discord about that.)

There are ways of avoiding creating merge commits in the first place, but I don't remember them off the top of my head.

**Pull request closed without merging**